### PR TITLE
Aleo Development Server

### PR DESCRIPTION
**Pull request by iamalwaysuncomfortable**
## Motivation

Most developers working building applications on Aleo are working with javascript or typescript environments. Unfortunately Aleo tooling mostly consists of rust libraries or cli tools which force developers to build somewhat painful bridges such as OS pipes within cloud containers in order to interact with the Aleo Blockchain.

The Aleo development server is meant to provide a simple REST API for use in local development or cloud-based applications.

## Test Plan
* Integration tests will be written against the server once its running that validate its creating successful executions and deployments.

### CODE REVIEW

File: rust/develop/README.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+#### Configuring Encrypted Private Keys
+The server can be started with an encrypted private key. If this option is used, the server
+will look for a password field in the body of incoming requests. If a password is provided
+(and the password is correct), the server will decrypt the private key and use it to build
+and send the transaction to the network. This mode of operation can be invoked with:
+
+`aleo-develop start --key-ciphertext <encrypted_private_key>`
+
+## Usage
+Once started, the he endpoints have the following options. All requests should be sent as a POST request with a json body.
```
Comment (collinc97) [inline rust/develop/README.md]: nit
```suggestion
Once started, the endpoints have the following options. All requests should be sent as a POST request with a json body.
```

File: rust/develop/README.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+#### Javascript API
+A javascript client for this server is available in the [Aleo SDK](https://www.npmjs.com/package/@aleohq/sdk)
+
+#### Endpoints
+`\develop`
+* `program`: Text representation of the program to be deployed
+* `fee`: Fee to be paid for the program deployment (REQUIRED)
+* `private_key`: Optional private key of the user who is deploying the program
```
Comment (collinc97) [inline rust/develop/README.md]: Make "Required" and "Optional" parameters consistent.
@iamalwaysuncomfortable if you have a preferred way to write these that is better than this suggestion that is fine.
```suggestion
* `fee`: Required fee to be paid for the program deployment
* `private_key`: Optional private key of the user who is deploying the program
```

File: rust/develop/README.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+* `password`: If the development server is started with an encrypted private key, the password will decrypt the private key
+* `fee_record`: Optional record in text format to be used for the fee. If not provided, the server will search the network for a suitable record to pay the fee.
+* `returns`: The transaction_id of the deployment transaction if successful
```
Comment (collinc97) [inline rust/develop/README.md]:
```suggestion
* `returns`: The transaction ID of the deployment transaction if successful
```

File: rust/develop/README.md [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+`\execute`
+* `program_id` The program_id of the program to be executed (e.g. hello.aleo)
```
Comment (collinc97) [inline rust/develop/README.md]:
```suggestion
* `program_id` The program ID of the program to be executed (e.g. hello.aleo)
```

File: rust/develop/README.md [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+* `program_function` The function to execute within the program (e.g. main)
+* `fee` Optional Fee to be paid for the transfer, specify 0 for no fee
```
Comment (collinc97) [inline rust/develop/README.md]:
```suggestion
* `fee` Optional fee to be paid for the transfer, specify 0 for no fee
```

File: rust/develop/README.md [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+* `inputs` Array of inputs to be passed to the program
+* `private_key` Optional private key of the user who is executing the program
+* `password`: If the development server is started with an encrypted private key, the password will decrypt the private key
+* `fee_record`: Optional record in text format to be used for the fee. If not provided, the server will search the network for a suitable record to pay the fee
+* `returns`: The transaction_id of the execution transaction if successful
```
Comment (collinc97) [inline rust/develop/README.md]:
```suggestion
* `returns`: The transaction ID of the execution transaction if successful
```

File: rust/develop/README.md [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+`\transfer`
+* `amount` The amount of credits to be sent (e.g. 1.5)
+* `fee` Optional Fee to be paid for the transfer, specify 0 for no fee
+* `recipient` The recipient of the transfer
+* `privateKey` Optional private key of the user who is sending the transfer
+* `password`: If the development server is started with an encrypted private key, the password will decrypt the private key
+* `amount_record` Optional record in text format to be used to fund the transfer. If not provided, the server will search the network for a suitable record to fund the amount
+* `fee_record` Optional record in text format to be used for the fee. If not provided, the server will search the network for a suitable record to pay the fee
+@returns {string | Error} The transaction_id of the execution transaction if successful
```
Comment (collinc97) [inline rust/develop/README.md]:
```suggestion
@returns {string | Error} The transaction ID of the execution transaction if successful
```

File: rust/develop/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+//! ## Usage
+//! Once started, the endpoints have the following options. All requests should be sent as a POST request with a json body.
+//!
+//! #### Javascript API
+//! A javascript client for this server is available in the [Aleo SDK](https://www.npmjs.com/package/@aleohq/sdk)
+//!
+//! #### Endpoints
+//! `\develop`
+//! * `program`: Text representation of the program to be deployed
+//! * `fee`: Fee to be paid for the program deployment (REQUIRED)
```
Comment (collinc97) [inline rust/develop/src/lib.rs]:
```suggestion
//! * `fee`: Required fee to be paid for the program deployment
```

File: rust/develop/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+//! * `private_key`: Optional private key of the user who is deploying the program
+//! * `password`: If the development server is started with an encrypted private key, the password will decrypt the private key
+//! * `fee_record`: Optional record in text format to be used for the fee. If not provided, the server will search the network for a suitable record to pay the fee.
+//! * `returns`: The transaction_id of the deployment transaction if successful
```
Comment (collinc97) [inline rust/develop/src/lib.rs]:
```suggestion
//! * `returns`: The transaction ID of the deployment transaction if successful
```

File: rust/develop/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+//!
+//! `\execute`
+//! * `program_id` The program_id of the program to be executed (e.g. hello.aleo)
```
Comment (collinc97) [inline rust/develop/src/lib.rs]:
```suggestion
//! * `program_id` The program ID of the program to be executed (e.g. hello.aleo)
```

File: rust/develop/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+//! * `program_function` The function to execute within the program (e.g. main)
+//! * `fee` Optional Fee to be paid for the transfer, specify 0 for no fee
```
Comment (collinc97) [inline rust/develop/src/lib.rs]:
```suggestion
//! * `fee` Optional fee to be paid for the transfer, specify 0 for no fee
```

File: rust/develop/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+//! * `inputs` Array of inputs to be passed to the program
+//! * `private_key` Optional private key of the user who is executing the program
+//! * `password`: If the development server is started with an encrypted private key, the password will decrypt the private key
+//! * `fee_record`: Optional record in text format to be used for the fee. If not provided, the server will search the network for a suitable record to pay the fee
+//! * `returns`: The transaction_id of the execution transaction if successful
```
Comment (collinc97) [inline rust/develop/src/lib.rs]:
```suggestion
//! * `returns`: The transaction ID of the execution transaction if successful
```

File: rust/develop/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+//!
+//! `\transfer`
+//! * `amount` The amount of credits to be sent (e.g. 1.5)
+//! * `fee` Optional Fee to be paid for the transfer, specify 0 for no fee
```
Comment (collinc97) [inline rust/develop/src/lib.rs]:
```suggestion
//! * `fee` Optional fee to be paid for the transfer, specify 0 for no fee
```

File: rust/develop/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+//! * `recipient` The recipient of the transfer
+//! * `privateKey` Optional private key of the user who is sending the transfer
+//! * `password`: If the development server is started with an encrypted private key, the password will decrypt the private key
+//! * `amount_record` Optional record in text format to be used to fund the transfer. If not provided, the server will search the network for a suitable record to fund the amount
+//! * `fee_record` Optional record in text format to be used for the fee. If not provided, the server will search the network for a suitable record to pay the fee
+//! @returns {string | Error} The transaction_id of the execution transaction if successful
```
Comment (collinc97) [inline rust/develop/src/lib.rs]:
```suggestion
//! @returns {string | Error} The transaction ID of the execution transaction if successful
```

**Review by collinc97 — CHANGES_REQUESTED** [review]
First pass over `rust/develop/` server.
Looking great - just a few english nits.

File: sdk/src/node_connection.js [hunk 15]
```diff
@@ -1,19 +1,20 @@
...
+import { RecordCiphertext, PrivateKey } from "@aleohq/wasm";
 /**
  * Connection management class that encapsulates REST calls to publicly exposed endpoints of Aleo nodes.
  * The methods provided in this class provide information on the Aleo Blockchain
  *
  * @param {string} host
  * @example
  * // Connection to a local node
- * let local_connection = new NodeConnection("localhost:4130");
+ * let local_connection = new NodeConnection("http://localhost:3030");
```
Comment (collinc97) [inline sdk/src/node_connection.js]:
```suggestion
 * let local_connection = new AleoNetworkClient("http://localhost:3030");
```

File: sdk/src/node_connection.js [hunk 16]
```diff
@@ -1,19 +1,20 @@
...
  *
  * // Connection to a public beacon node
- * let public_connection = new NodeConnection("vm.aleo.org/api");
+ * let public_connection = new NodeConnection("https://vm.aleo.org/api");
```
Comment (collinc97) [inline sdk/src/node_connection.js]:
```suggestion
 * let public_connection = new AleoNetworkClient("https://vm.aleo.org/api");
```

File: sdk/src/node_connection.js
Comment (collinc97) [inline sdk/src/node_connection.js:1]: Should this file be renamed to `aleo_network_client.js`?

**Review by collinc97 — CHANGES_REQUESTED** [review]
This functionality is vital to making Aleo web-ready.
Thanks for the hard work @iamalwaysuncomfortable!

I think the naming between `AleoNetworkClient` and `DevelopmentClient` can be improved.
In my opinion, they should both be `AleoNetworkClient` and `AleoDevelopmentClient`  
or `NetworkClient` and `DevelopmentClient`  
or maybe something more clear.
Let's discuss during the next sync.

Code-wise this PR looks good!

File: sdk/src/node_connection.js (hunk 16 above)
Comment (iamalwaysuncomfortable) [reply sdk/src/node_connection.js]: Good suggestion, done!

File: sdk/src/node_connection.js
Comment (iamalwaysuncomfortable) [reply sdk/src/node_connection.js:1]: Yes, done!

File: sdk/src/node_connection.js (hunk 15 above)
Comment (iamalwaysuncomfortable) [reply sdk/src/node_connection.js]: Absolutely agree

**Review by collinc97 — APPROVED** [review]

**Pull request merged**